moveit_ros/planning/moveit_cpp/src/planning_component.cpp

```diff
@@ -189,7 +216,127 @@ PlanningComponent::PlanSolution PlanningComponent::plan(const PlanRequestParamet
   //    visual_tools_->publishRobotState(last_solution_trajectory_->getLastWayPoint(), rviz_visual_tools::TRANSLUCENT);
   //  }
   //}
-  return *last_plan_solution_;
+
+  if (update_last_solution)
+  {
+    last_plan_solution_ = plan_solution;
+  }
+  return *plan_solution;
+}
+
+PlanningComponent::PlanSolution PlanningComponent::plan(const MultiPipelinePlanRequestParameters& parameters,
+                                                        SolutionCallbackFunction solution_selection_callback,
+                                                        StoppingCriterionFunction stopping_criterion_callback)
+{
+  // Create solutions container
+  PlanningComponent::PlanSolutions planning_solutions{ parameters.multi_plan_request_parameters.size() };
+  std::vector<std::thread> planning_threads;
+  planning_threads.reserve(parameters.multi_plan_request_parameters.size());
+
+  // Launch planning threads
+  for (auto const& plan_request_parameter : parameters.multi_plan_request_parameters)
+  {
+    auto planning_thread = std::thread([&]() {
+      try
+      {
+        auto solution = plan(plan_request_parameter, false);
+        planning_solutions.pushBack(solution);
+      }
+      catch (const std::exception& e)
+      {
+        RCLCPP_ERROR_STREAM(LOGGER, "Planning pipeline '" << plan_request_parameter.planning_pipeline.c_str()
+                                                          << "' threw exception '" << e.what() << "'");
+        auto plan_solution = std::make_shared<PlanSolution>();
+        plan_solution->error_code = moveit::core::MoveItErrorCode::FAILURE;
+      }
+    });
+    planning_threads.push_back(std::move(planning_thread));
+  }
+
+  // Launch monitor thread if a stopping criterion is provided as function arguments. Otherwise,
+  // the parallel planning process takes as long as the longest planner takes.
+  if (stopping_criterion_callback != nullptr)
+  {
+    auto monitor_thread = std::thread([&]() {
+      while (planning_solutions.getSolutions().size() < planning_threads.size())
+      {
+        if (stopping_criterion_callback(planning_solutions, parameters))
```

Comment:
It wouldn't be too bad to ensure that the stopping criterion is thread-safe. Also, if the callback is dependent on solutions, I would simply check it when a solution is returned in the planning thread. A time-out can be handled explicitly instead of this busy-wait loop

Reply:
That is a good idea. This way I can completely remove the monitor thread :+1: I removed the load() function from the MultiPipelinePlanRequestParameters struct so the StoppingCriterionFunction should be thread safe now:
```cpp
  typedef std::function<bool(PlanSolutions const& solutions,
                             MultiPipelinePlanRequestParameters const& plan_request_parameters)>
      StoppingCriterionFunction;
```